File: apps/api/internal/domain/workout.go
package domain

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/datatypes"
)

type WorkoutPlan struct {
	ID                       uuid.UUID             `gorm:"type:uuid;primary_key" json:"id"`
	UserID                   uuid.UUID             `gorm:"type:uuid;not null" json:"-"`
	Name                     string                `gorm:"type:text;not null" json:"title"`
	Description              *string               `gorm:"type:text" json:"description"`
	Difficulty               *string               `gorm:"type:varchar(50)" json:"difficulty"`
	EstimatedDurationMinutes *int16                `gorm:"type:smallint" json:"estimatedDurationMinutes"`
	GeneratedByAI            bool                  `gorm:"not null;default:false" json:"generatedByAI"`
	AIMetadata               datatypes.JSON        `gorm:"type:jsonb" json:"aiMetadata"`
	CreatedAt                time.Time             `gorm:"not null;default:current_timestamp" json:"createdAt"`
	UpdatedAt                time.Time             `gorm:"not null;default:current_timestamp" json:"updatedAt"`
	DeletedAt                *time.Time            `json:"-"`
	Exercises                []WorkoutPlanExercise `gorm:"foreignKey:WorkoutPlanID" json:"exercises"`
}

type WorkoutPlanExercise struct {
	ID                    uuid.UUID `gorm:"type:uuid;primary_key" json:"id"`
	WorkoutPlanID         uuid.UUID `gorm:"type:uuid;not null" json:"-"`
	ExerciseID            uuid.UUID `gorm:"type:uuid;not null" json:"exerciseId"`
	OrderIndex            int16     `gorm:"type:smallint;not null" json:"orderIndex"`
	TargetSets            int16     `gorm:"type:smallint;not null" json:"targetSets"`
	TargetReps            *int16    `gorm:"type:smallint" json:"targetReps"`
	TargetDurationSeconds *int16    `gorm:"type:smallint" json:"targetDurationSeconds"`
	TargetRestSeconds     *int16    `gorm:"type:smallint" json:"restSeconds"`
	CreatedAt             time.Time `gorm:"not null;default:current_timestamp" json:"createdAt"`
	Exercise              Exercise  `gorm:"foreignKey:ExerciseID" json:"exercise"`
}

type WorkoutSession struct {
	ID             uuid.UUID       `gorm:"type:uuid;primary_key" json:"id"`
	UserID         uuid.UUID       `gorm:"type:uuid;not null" json:"-"`
	WorkoutPlanID  *uuid.UUID      `gorm:"type:uuid" json:"planId"`
	ClientEventID  uuid.UUID       `gorm:"type:uuid;uniqueIndex;not null" json:"clientEventId"`
	StartTime      time.Time       `gorm:"not null" json:"startTime"`
	EndTime        *time.Time      `json:"endTime"`
	Status         string          `gorm:"type:varchar(50);not null;default:'in_progress'" json:"status"`
	CaloriesBurned *float64        `gorm:"type:numeric(6,2)" json:"caloriesBurned"`
	CreatedAt      time.Time       `gorm:"not null;default:current_timestamp" json:"createdAt"`
	UpdatedAt      time.Time       `gorm:"not null;default:current_timestamp" json:"updatedAt"`
	Logs           []WorkoutSetLog `gorm:"foreignKey:SessionID" json:"logs"`
}

type WorkoutSetLog struct {
	ID              uuid.UUID      `gorm:"type:uuid;primary_key" json:"id"`
	SessionID       uuid.UUID      `gorm:"type:uuid;not null" json:"-"`
	ExerciseID      uuid.UUID      `gorm:"type:uuid;not null" json:"exerciseId"`
	SetNumber       int16          `gorm:"type:smallint;not null" json:"setNumber"`
	Reps            *int16         `gorm:"type:smallint" json:"reps"`
	WeightKg        *float64       `gorm:"type:numeric(6,2)" json:"weightKg"`
	DurationSeconds *int16         `gorm:"type:smallint" json:"durationSeconds"`
	AIPoseFeedback  datatypes.JSON `gorm:"type:jsonb" json:"aiPoseFeedback"`
	CreatedAt       time.Time      `gorm:"not null;default:current_timestamp" json:"createdAt"`
}

File: apps/mobile/src/models/workout.ts
import { z } from 'zod';

export const WorkoutExerciseSchema = z.object({
  id: z.string(),
  name: z.string(),
  imageUrl: z.string().optional(),
  videoUrl: z.string().optional(),
  instructions: z.array(z.string()).default([]),
  targetSets: z.number(),
  targetReps: z.number().optional().nullable(),
  targetDurationSeconds: z.number().optional().nullable(),
  restSeconds: z.number().default(60),
  equipment: z.string().optional(),
  muscleGroup: z.string().optional(),
});
export type WorkoutExercise = z.infer<typeof WorkoutExerciseSchema>;

export const WorkoutPlanSchema = z.object({
  id: z.string(),
  title: z.string(),
  description: z.string().optional(),
  difficulty: z.enum(['beginner', 'intermediate', 'advanced']).default('intermediate'),
  estimatedDurationMinutes: z.number().default(45),
  estimatedCalories: z.number().optional().nullable(),
  equipment: z.array(z.string()).default([]),
  targetMuscles: z.array(z.string()).default([]),
  generatedByAI: z.boolean().default(false),
  exercises: z.array(WorkoutExerciseSchema).default([]),
});
export type WorkoutPlan = z.infer<typeof WorkoutPlanSchema>;

export const WorkoutSetLogSchema = z.object({
  id: z.string().optional(),
  exerciseId: z.string(),
  setNumber: z.number(),
  reps: z.number().optional().nullable(),
  weightKg: z.number().optional().nullable(),
  durationSeconds: z.number().optional().nullable(),
});
export type WorkoutSetLog = z.infer<typeof WorkoutSetLogSchema>;

export const WorkoutSessionSchema = z.object({
  id: z.string(),
  clientEventId: z.string(),
  planId: z.string().optional().nullable(),
  startTime: z.string(),
  endTime: z.string().optional().nullable(),
  status: z.string(),
  caloriesBurned: z.number().optional().nullable(),
  logs: z.array(WorkoutSetLogSchema).default([]),
});
export type WorkoutSession = z.infer<typeof WorkoutSessionSchema>;
